crates/key-value-aws/src/store.rs

```diff
@@ -0,0 +1,499 @@
+            .table_name(self.table.as_str())
+            .key(PK, AttributeValue::S(self.key.clone()))
+            .update_expression("SET #lock=:lock")
+            .expression_attribute_names("#lock", LOCK)
+            .expression_attribute_values(":lock", AttributeValue::Null(true))
+            .condition_expression("attribute_not_exists (#lock)")
+            .return_values(aws_sdk_dynamodb::types::ReturnValue::AllNew)
+            .send()
+            .await
+            .map_err(log_error)?;
+
+        self.has_lock.lock().unwrap().clone_from(&true);
+
+        match attributes {
+            Some(mut item) => match item.remove(VAL) {
+                Some(AttributeValue::B(val)) => Ok(Some(val.into_inner())),
+                _ => Ok(None),
+            },
+            None => Ok(None),
+        }
+    }
+
+    /// `swap` updates the value for the key using the version saved in the `current` function for
+    /// optimistic concurrency.
+    async fn swap(&self, value: Vec<u8>) -> Result<(), SwapError> {
+        let mut update_item = Update::builder()
+            .table_name(self.table.as_str())
+            .key(PK, AttributeValue::S(self.key.clone()))
+            .update_expression("SET #val=:val REMOVE #lock")
+            .expression_attribute_names("#val", VAL)
+            .expression_attribute_values(":val", AttributeValue::B(Blob::new(value)))
+            .expression_attribute_names("#lock", LOCK);
+
+        let has_lock = *self.has_lock.lock().unwrap();
+        // Ensure exclusive access between fetching the current value of the item and swapping
+        if has_lock {
+            update_item = update_item.condition_expression("attribute_exists (#lock)");
+        }
+
+        // TransactWriteItems fails if concurrent writes are in progress on an item, so even without locking, we get atomicity in overwriting
+        self.client
+            .transact_write_items()
+            .transact_items(
+                TransactWriteItem::builder()
+                    .update(
+                        update_item
+                            .build()
+                            .map_err(|e| SwapError::Other(format!("{e:?}")))?,
+                    )
+                    .build(),
+            )
+            .send()
+            .await
+            .map_err(|e| SwapError::CasFailed(format!("{e:?}")))?;
+
+        Ok(())
+    }
+
+    async fn bucket_rep(&self) -> u32 {
+        self.bucket_rep
+    }
+
+    async fn key(&self) -> String {
+        self.key.clone()
+    }
+}
+
+impl AwsDynamoStore {
+    async fn get_keys(&self) -> Result<Vec<String>, Error> {
+        let mut primary_keys = Vec::new();
+        let mut last_evaluated_key = None;
+
+        loop {
+            let mut scan_builder = self
+                .client
+                .scan()
+                .table_name(self.table.as_str())
+                .projection_expression(PK);
+
+            if let Some(keys) = last_evaluated_key {
```

Comment:
This would potentially be slightly easier to read with the SDK's paginator (https://docs.rs/aws-sdk-dynamodb/latest/aws_sdk_dynamodb/operation/scan/builders/struct.ScanFluentBuilder.html#method.into_paginator) - but I'm extremely not a Rustacean 😅  (and this otherwise is doing ~the same thing afaict)

Reply:
Good point -- I was unaware of this utility! I will look into replacing custom logic with the paginator.